nexus/src/authn/external/mod.rs

```diff
@@ -35,13 +36,16 @@ where
     // TODO-openapi: At some point, the authentication headers need to get into
     // the OpenAPI spec.  We probably don't want to have every endpoint function
     // accept them via an extractor, though.
-    pub async fn authn_request(
+    pub async fn authn_request<Q>(
         &self,
-        rqctx: &dropshot::RequestContext<T>,
-    ) -> Result<authn::Context, authn::Error> {
+        rqctx: &dropshot::RequestContext<Q>,
+    ) -> Result<authn::Context, authn::Error>
+    where
+        Q: Borrow<T> + Send + Sync + 'static,
```

Comment:
I think we want 

```suggestion
        Q: AsRef<T> + Send + Sync + 'static,
```

(and `.as_ref()` in the function body, accordingly). From [the first edition of the Rust book](https://doc.rust-lang.org/1.30.0/book/first-edition/borrow-and-asref.html):

> Which should I use?
>
> We can see how they’re kind of the same: they both deal with owned and borrowed versions of some type. However, they’re a bit different.
> 
> Choose Borrow when you want to abstract over different kinds of borrowing, or when you’re building a data structure that treats owned and borrowed values in equivalent ways, such as hashing and comparison.
>
> Choose AsRef when you want to convert something to a reference directly, and you’re writing generic code.

I think we're in the latter case here.

Reply:
We chatted about this offline - even though all the non-test code always has `RequestContext<Arc<Value>>` (which would work with `AsRef`), there's [a unit test](https://github.com/oxidecomputer/omicron/blob/c2d033c4f8b61c8fc9399098c2cefb9c4ddb720d/nexus/tests/integration_tests/authn_http.rs#L402) that has a `dropshot::RequestContext<Value>` without an `Arc`. `AsRef<T>` is not implemented for `&T`, so we're back in this case for `Borrow`:

> when you’re building a data structure that treats owned and borrowed values in equivalent ways

Reply:
To add to that: we could also change the test case to use `Arc`.  But it feels (to me anyway) that yeah, this interface doesn't care if the data is owned or borrowed -- the caller shouldn't have to worry about this.  Thanks for digging in!